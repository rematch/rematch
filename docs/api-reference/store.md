---
id: store
title: Store
sidebar_label: Store
slug: /api-reference/store
---

Store returned by Rematch is essentially a [Redux store](https://redux.js.org/api/store) with a few additional properties and extra features. Therefore, for more in-depth reference of the functions '_provided by Redux_', refer to the Redux documentation.

---

### Configuration

Store is an object which contains the following properties:

- `name` (_string_): name of a store

- `addModel(namedModel)`: it allows to lazy-load models and merge them into Rematch after [init](/docs/api-reference#initconfig) has been called. **addModel** accepts _namedModel_ which means it needs to have parameter `name` defined (see example below).

- `getState()`: provided by Redux, returns the state of your store.

- `subscribe(listener: () => void): Unsubscribe`: provided by Redux, adds a change listener.

- `replaceReducer(nextReducer): void`: provided by Redux, replaces the reducer currently used by the store to calculate the state.

- `dispatch(action)`: provided by Redux, a function that dispatches actions to the store. However, in Rematch **dispatch** can be also used as an object which contains action dispatchers (see example below).

### Dispatch actions directly from store

```js
import store from "./store";

// dispatch reducers actions
store.dispatch({ type: "count/increment", payload: 1 }); // regular dispatch usage
store.dispatch.count.increment(1); // the same as above but with action dispatcher

// dispatch effects actions
store.dispatch({ type: "count/incrementAsync", payload: 1 }); // regular dispatch usage
store.dispatch.count.incrementAsync(1); // the same as above but with action dispatcher
```

### Add dynamic models

```js
import store from "./store";
// (1) initially
store.getState(); // { count: 0 }
// (2) add model
store.addModel({ name: "countB", state: 99 });
// (3) after adding
store.getState(); // { count: 0, countB: 99 }
```
