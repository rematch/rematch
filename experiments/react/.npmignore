# directories
src
plugins
packages
flow-typed
test
examples
scripts
.vscode
.coverage

# files
.babelrc
.gitignore
.prettierrc
.travis.yml
webpack.config.js
yarn.lock
