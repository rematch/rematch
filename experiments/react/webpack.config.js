/* global __dirname, require, module */

const webpack = require('webpack')
const path = require('path')
const libraryName = require('./package.json').name
const { env } = require('yargs').argv

const { UglifyJsPlugin } = webpack.optimize

const plugins = [
  new webpack.DefinePlugin({
    'process.env': {
      NODE_ENV: JSON.stringify('production')
    }
  })
]
let outputFile

if (env === 'build') {
  plugins.push(new UglifyJsPlugin({ minimize: true }))
  outputFile = 'index.min.js'
} else {
  outputFile = 'index.js'
}

const config = {
  entry: [`${__dirname}/src/index.js`],
  devtool: 'source-map',
  output: {
    path: `${__dirname}/lib`,
    filename: outputFile,
    library: libraryName,
    libraryTarget: 'umd',
  },
  module: {
    rules: [
      {
        test: /(\.jsx|\.js)$/,
        loader: 'babel-loader',
        exclude: /(node_modules)/,
      },
    ],
  },
  resolve: {
    modules: [path.resolve('./node_modules'), path.resolve('./src')],
    extensions: ['.json', '.js'],
  },
  plugins,
}

module.exports = config
