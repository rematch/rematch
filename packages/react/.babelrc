{
  "presets": ["react", "env", "flow"],
  "plugins": [
    "transform-flow-strip-types",
    "transform-object-rest-spread"
  ]
}