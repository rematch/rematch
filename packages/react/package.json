{
  "name": "@rematch/react",
  "version": "0.1.0-alpha.1",
  "description": "React bindings for Rematch",
  "keywords": [
    "@rematch",
    "redux",
    "rematch",
    "react"
  ],
  "homepage": "https://github.com/rematch/rematch/tree/master/packages/react#readme",
  "bugs": {
    "url": "git+https://github.com/rematch/rematch/issues"
  },
  "license": "ISC",
  "author": "Blair bodnar <blairbodnar@gmail.com>",
  "files": [
    "lib"
  ],
  "main": "lib/index.min.js",
  "scripts": {
    "build": "webpack --env build"
  },
  "dependencies": {
    "react": "^16.1.0",
    "react-redux": "^5.0.6",
    "@rematch/core": "^0.1.0-alpha.13"
  },
  "devDependencies": {
    "babel-core": "^6.26.0",
    "babel-eslint": "^8.0.2",
    "babel-loader": "^7.1.2",
    "babel-plugin-transform-flow-strip-types": "^6.22.0",
    "babel-plugin-transform-object-rest-spread": "6.26.0",
    "babel-plugin-transform-runtime": "^6.23.0",
    "babel-preset-env": "^1.6.1",
    "babel-preset-flow": "^6.23.0",
    "babel-preset-react": "^6.24.1",
    "flow-bin": "^0.59.0",
    "jest": "^21.2.1",
    "webpack": "^3.8.1",
    "yargs": "^10.0.3"
  }
}
