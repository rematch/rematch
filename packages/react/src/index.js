import React from 'react'
import { connect, Provider } from 'react-redux'

let dispatch = null
let select = null
let getStore = null

const useReact = Rematch => {
  dispatch = Rematch.dispatch
  select = Rematch.select
  getStore = Rematch.getStore
}

const getModelSelectors = (state, modelName) => {
  const modelSelectors = {}
  Object.keys(select[modelName] || {}).forEach(selectorName => {
    modelSelectors[selectorName] = (...args) => select[modelName][selectorName](state, ...args)
  })
  return modelSelectors
}

const connectRematch = (modelNames, mapToProps) => Component => {
  const processModels = state => {
    const models = {}
    modelNames.forEach(name => {
      models[name] = {
        name,
        state: state[name],
        dispatch: dispatch[name],
        select: getModelSelectors(state, name)
      }
    })
    return models
  }
  return connect(state => mapToProps(processModels(state)))(Component)
}

const RematchReact = props => (
  <Provider store={getStore()}>
    {props.children}
  </Provider>
)

export {
  useReact,
  RematchReact,
  connectRematch as connect
}

