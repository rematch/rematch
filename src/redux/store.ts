/* eslint no-underscore-dangle: 0 */
import { applyMiddleware, createStore as _createStore, Middleware, Reducer, Store, StoreCreator } from 'redux'
import { Config, MiddlewareDefinition, Model, RootReducers } from '../../typings/rematch'
import CoreFactory from '../core'
import { composeEnhancers } from './devtools'
import ReducersFactory from './reducers'

export default class LocalStore<S> {

  public readonly pluginMiddlewares: Array<MiddlewareDefinition<S>> = []

  public store: Store<S>
  private rootReducers: RootReducers
  private reducersFactory: ReducersFactory<S>

  constructor(reducersFactory: ReducersFactory<S>) {
    this.reducersFactory = reducersFactory
  }

  public initStore = ({ redux }: Config<S>): Store<S> => {
    const initialState: any = typeof redux.initialState === 'undefined' ? {} : redux.initialState
    const createStore: StoreCreator = redux.createStore || _createStore
    this.rootReducers = redux.rootReducers
    const rootReducer: Reducer<any> = this.reducersFactory.createRootReducer(this.rootReducers)
    const middlewareList: Array<MiddlewareDefinition<S>> =
      [...this.pluginMiddlewares, ...(redux.middlewares || [])]
    const middlewares = applyMiddleware(...middlewareList as Middleware[]) // FIXME: Not sure how to avoid this
    const enhancers = composeEnhancers(redux.devtoolOptions)(...redux.enhancers || [], middlewares)
    this.store = createStore(rootReducer, initialState, enhancers)
    return this.store
  }

  // allows for "model" to dynamically update the reducers/store
  // public createReducersAndUpdateStore = (model: Model<S>): void => {
  //   this.reducersFactory.mergeReducers(this.reducersFactory.createModelReducer(model))
  //   this.store.replaceReducer(this.reducersFactory.createRootReducer(this.rootReducers))
  // }

}
