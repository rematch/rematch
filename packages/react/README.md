# Rematch React

Better README coming soon... For now here's the example:

```jsx
import React from 'react'
import { render } from 'react-dom'
import Rematch from '@rematch/core'
import { useReact, connect, RematchReact } from '@rematch/react'

useReact(Rematch)

Rematch.init({
  models: {
    user: {
      name: 'user',
      state: { name: 'Blair', location: 'Vancouver' }
    },
    count: {
      name: 'count',
      state: 0,
      reducers: { up: state => state + 1 },
      selectors: { double: state => state * 2 }
    },
  }
})

const Component = ({user, count}) => (
  <ul>
    <li>Welcome, {user.state.name} from {user.state.location}</li>
    <li>The count is {count.state}</li>
    <li>The count doubled is {count.select.double()}</li>
    <li><button onClick={() => count.dispatch.up()}>count Up</button></li>
  </ul>
)

const Container = connect(['user', 'count'], ({user, count}) => ({
  user,
  count,
}))(Component)

render(
  <RematchReact>
    <Container />
  </RematchReact>,
  document.getElementById('root')
)
```