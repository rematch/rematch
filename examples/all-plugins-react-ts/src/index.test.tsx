import React from 'react'
import { render } from '@testing-library/react'
import { Provider } from 'react-redux'

import { GlobalApp } from './GlobalApp'
import { store } from './store'

const TestingProvider: React.FC = ({ children }) => (
	<Provider store={store}>{children}</Provider>
)

test('Application is rendered correctly', () => {
	const { container } = render(<GlobalApp />, { wrapper: TestingProvider })
	expect(container).toMatchSnapshot()
})

test('Store is correctly initialized', () => {
	render(<GlobalApp />, { wrapper: TestingProvider })
	expect(store.getState()).toEqual({
		_persist: {
			rehydrated: true,
			version: -1,
		},
		cart: {
			taxPercent: 8,
			items: [
				{ name: 'apple', value: 1.2 },
				{ name: 'orange', value: 0.95 },
			],
		},
		loading: {
			effects: {
				cart: {},
				players: {
					getPlayers: true,
				},
				settings: {},
				updated: {},
			},
			global: true,
			models: {
				cart: false,
				players: true,
				settings: false,
				updated: false,
			},
		},
		players: {
			players: [],
		},
		settings: {
			isLightThemeOn: true,
		},
		updated: {
			cart: {},
			loading: {},
			players: {},
			settings: {},
		},
	})
})
