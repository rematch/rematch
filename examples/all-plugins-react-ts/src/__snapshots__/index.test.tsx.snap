// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application is rendered correctly 1`] = `
<div>
  <div>
    <button
      type="button"
    >
      Show Application with hooks
    </button>
    <button
      type="button"
    >
      Show Application with classes
    </button>
    <div>
      <div
        style="display: flex; width: 100%; justify-content: space-between;"
      >
        <h1>
          NBA Hook Players:
        </h1>
        <div
          class="theme-switch-wrapper"
        >
          <label
            class="theme-switch"
          >
            <input
              id="checkbox"
              type="checkbox"
            />
            <div
              class="slider round"
            />
          </label>
          <em>
            Enable Dark Mode!
          </em>
        </div>
      </div>
      <div
        class="container"
      >
        <div
          class="loader"
        >
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
